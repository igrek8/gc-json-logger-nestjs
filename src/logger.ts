import { Logger as ContextLogger, ILogger, LogEntryMetadata, Severity } from 'gc-json-logger';

export class Logger implements ILogger {
  constructor(protected context?: string, protected labels: Record<string, string | undefined> = {}) {}

  log(severity: Severity, message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.log(severity, message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  default(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.default(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  debug(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.debug(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  info(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.info(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  notice(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.notice(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  warning(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.warning(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  error(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.error(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  critical(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.critical(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  alert(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.alert(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  emergency(message: string, meta?: LogEntryMetadata | undefined): void {
    ContextLogger.emergency(message, {
      ...meta,
      context: this.context,
      labels: { ...meta?.labels, ...this.labels },
    });
  }

  setLabels(labels: Record<string, string | undefined>): void {
    this.labels = labels;
  }

  getLabels(): Record<string, string | undefined> {
    return this.labels;
  }

  getName(): string {
    return this.context ?? ContextLogger.getName();
  }
}
