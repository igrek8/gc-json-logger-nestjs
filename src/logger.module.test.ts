import { Body, Controller, HttpCode, INestApplication, Inject, Post } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { Logger } from 'gc-json-logger';
import request from 'supertest';
import { LoggerModule } from './logger.module';
import { LoggerService } from './logger.service';

jest.useFakeTimers();
jest.setSystemTime(0);

@Controller()
class TestController {
  constructor(@Inject(LoggerService) protected readonly logger: LoggerService) {}

  @Post('/echo')
  @HttpCode(200)
  echo(@Body() echo: object) {
    this.logger.info('received data', { echo });
    return { ...echo };
  }
}

describe('LoggerModule', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        LoggerModule.register({
          configure: (middleware) => middleware.forRoutes('*'),
        }),
      ],
      controllers: [TestController],
    }).compile();
    app = moduleRef.createNestApplication();
    await app.listen(50000);
  });

  afterAll(async () => {
    await app.close();
  });

  it('logs a request', async () => {
    const log = jest.spyOn(Logger.prototype, 'log');
    log.mockImplementation(() => {});
    await request(app.getHttpServer())
      .post('/echo')
      .set('user-agent', 'curl/7.64.1')
      .set('referer', 'https://example.com/')
      .send({ message: 'test' });
    expect(log.mock.calls).toMatchSnapshot();
  });
});
