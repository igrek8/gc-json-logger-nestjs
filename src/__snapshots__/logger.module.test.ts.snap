// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoggerModule logs a request 1`] = `
[
  [
    200,
    "200 POST /",
    {
      "httpRequest": {
        "protocol": "HTTP/1.1",
        "referer": "https://example.com/",
        "remoteIp": "::ffff:127.0.0.1",
        "requestMethod": "POST",
        "requestSize": "223",
        "requestUrl": "http://127.0.0.1:50000/echo",
        "serverIp": "::ffff:127.0.0.1:50000",
        "userAgent": "curl/7.64.1",
      },
    },
  ],
  [
    200,
    "received data",
    {
      "echo": {
        "message": "test",
      },
    },
  ],
  [
    200,
    "200 POST /echo (0ms)",
    {
      "httpRequest": {
        "latency": "0ms",
        "responseSize": "225",
        "status": 200,
      },
      "request": {
        "body": {
          "message": "test",
        },
        "headers": {
          "accept-encoding": "gzip, deflate",
          "connection": "close",
          "content-length": "18",
          "content-type": "application/json",
          "host": "127.0.0.1:50000",
          "referer": "https://example.com/",
          "user-agent": "curl/7.64.1",
        },
      },
      "response": {
        "body": {
          "message": "test",
        },
        "headers": {
          "content-length": "18",
          "content-type": "application/json; charset=utf-8",
          "etag": "W/"12-o4TMqk657HDSoDIn46l7b2N/YX8"",
          "x-powered-by": "Express",
        },
      },
    },
  ],
]
`;
